planetiler-core/src/main/java/com/onthegomap/planetiler/files/WriteableFilesArchive.java

```diff
@@ -0,0 +1,250 @@
+      case ASYNC -> new AsyncTileFilesWriter(basePath, tileSchemeEncoder, bytesWritten);
+    };
+  }
+
+  @Override
+  public void finish(TileArchiveMetadata tileArchiveMetadata) {
+    if (metadataPath == null) {
+      return;
+    }
+    try (OutputStream s = new CountingOutputStream(Files.newOutputStream(metadataPath), bytesWritten::add)) {
+      TileArchiveMetadataDeSer.mbtilesMapper().writeValue(s, tileArchiveMetadata);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public long bytesWritten() {
+    return bytesWritten.sum();
+  }
+
+  @Override
+  public void close() throws IOException {
+    // nothing to do here
+  }
+
+  private static Path createValidateDirectory(Path p) {
+    if (!Files.exists(p)) {
+      FileUtils.createDirectory(p);
+    }
+    Preconditions.checkArgument(
+      Files.isDirectory(p),
+      "require \"" + p + "\" to be a directory"
+    );
+    return p;
+  }
+
+  private abstract static class BaseTileFilesWriter implements TileWriter {
+
+    private final Function<TileCoord, Path> tileSchemeEncoder;
+    private final LongAdder bytesWritten;
+    private Path lastCheckedFolder;
+
+    BaseTileFilesWriter(Path basePath, Function<TileCoord, Path> tileSchemeEncoder, LongAdder bytesWritten) {
+      this.tileSchemeEncoder = tileSchemeEncoder;
+      this.lastCheckedFolder = basePath;
+      this.bytesWritten = bytesWritten;
+    }
+
+    @Override
+    public final void write(TileEncodingResult encodingResult) {
+
+      final byte[] data = encodingResult.tileData();
+
+      final Path file = tileSchemeEncoder.apply(encodingResult.coord());
+      final Path folder = file.getParent();
+
+      // tiny optimization in order to avoid too many unnecessary "folder-exists-checks" (I/O)
+      // only effective when the tileScheme is z/x/y but doesn't really harm otherwise
+      if (!lastCheckedFolder.equals(folder) && !Files.exists(folder)) {
+        FileUtils.createDirectory(folder);
+      }
+      lastCheckedFolder = folder;
+      try {
+        writeData(file, data);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+
+      bytesWritten.add(data.length);
+    }
+
+    protected abstract void writeData(Path p, byte[] b) throws IOException;
+  }
+
+  private static class SyncTileFilesWriter extends BaseTileFilesWriter {
+
+    SyncTileFilesWriter(Path basePath, Function<TileCoord, Path> tileSchemeEncoder, LongAdder bytesWritten) {
+      super(basePath, tileSchemeEncoder, bytesWritten);
+    }
+
+    @Override
+    protected void writeData(Path p, byte[] b) throws IOException {
+      try {
+        Files.write(p, b);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
+    @Override
+    public void close() {
+      // nothing to do here
+    }
+  }
+
+  private static class AsyncTileFilesWriter extends BaseTileFilesWriter {
```

Comment:
Let's remove the async tile writer and collapse into a single sync writer class here for simplicity